src/algorithms/cf.py

```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
+    """
+
+    # add probR to [(user_id, item_id, rating)]
+    components = rCombo_ui.join(probR)
+
+    # add probRU to [(user_id, item_id, rating, probR)]
+    tmp_a = components.map(lambda (r, ((u,i), prob_r)): ((u, r), (i, prob_r)))
+    tmp_b = probRU.map(lambda (u, r, prob_ru): ((u, r), prob_ru))
+    components = tmp_a.join(tmp_b)
+
+    # add probRI to [(user_id, item_id, rating, probR, probRU)]
+    tmp_a = components.map(lambda ((u, r), ((i, prob_r), prob_ru)): (i, r), (u, prob_r, prob_ru))
+    tmp_b = probRI.map(lambda (i, r, prob_ri): ((i, r), prob_ri))
+    components = tmp_a.join(tmp_b)
+
+    # re-format
+    # [((user_id, movie_id, rating), naive_bayes_probability)]
+    componentsReformat = components.map(lambda ((i, r), ((u, prob_r, prob_ru), prob_ri)): ((u,i,r), (prob_r, prob_ru, prob_ri)))
+    bayesProb = componentsReformat.mapValues(lambda (prob_r, prob_ru, prob_ri): prob_ru * prob_ri / prob_r)
+
+    # [((user_id, item_id), [(rating_1, bayes_prob_1), (rating_2, bayes_prob_2), ..., (rating_5, bayes_prob_5)])]
+    # sort it by the lowest to the highest rating
+    ui_allBayesProb = bayesProb.mapValues(lambda value: [value]).reduceByKey(lambda a, b: a + b)\
+                               .mapValues(lambda value: sorted(value, key=lambda(rating, bayes_prob): rating))
+
+    return ui_allBayesProb
+
+def calc_naive_bayes_map(training_data, sc, computeFromScratch=True, ui_allBayesProb=None):
+    """
+    Determine the predicted rating of every user-item combination using Naive Bayes MAP.
+    Pai     : predicted rating for user a on item i
+    P(r|a,i): Naive Bayes that computes the probability of rating r for a given user a on item i
+    Pai = Argmax(r=1 to 5) P(r|a,i)
+
+    Assumption:
+    Since Naive Bayes can be defined as P(r|u,i) = ( (P(r|u) * P(r|i)) / P(r) ) * ( (P(u) * P(i)) / P(u, i)), 
+    we make the assumption that the latter part of the multiplication, ( (P(u) * P(i)) / P(u, i)), can be ignored.
+
+    Args:
+        training_data: the data used to train the RecSys algorithm in the format of a RDD of [ (userId, itemId, actualRating) ]
+
+    Returns:
+        predictions: predicted ratings of every user-item combination in the format of a RDD of [(userId, itemId, predictedRating)].
+    """
+
+    def calculate_bayes_map(value):
+        # extract the bayes_prob 
+        bayesProbList = [x[1] for x in value]
+        
+        # define the argmax, return the index
+        argmax = bayesProbList.index(max(bayesProbList))
+        
+        return argmax
+
+    if computeFromScratch:
+        ui_allBayesProb = __calc_naive_bayes_components(training_data, sc)
+    else:
+        if ui_allBayesProb is None:
+            raise Exception("ERROR: ui_allBayesProb is not defined although user specified not to calculate ui_allBayesProb not from scratch.")
+
+    return ui_allBayesProb.mapValues(calculate_bayes_map)
+
+
+def calc_naive_bayes_mse(training_data, sc):
```

Comment:
Noted that it requires additional arguments: (training_data, sc, computeFromScratch=True, ui_allBayesProb=None)